omero_figure/scripts/omero/figure_scripts/Figure_To_Pdf.py

```diff
@@ -1038,6 +1038,133 @@ def draw_scalebar(self, panel, region_width, page):
                 label, (lx + lx_end)/2, ly, font_size, (red, green, blue),
                 align="center")
 
+    def render_big_image_region(self, image, z, t, region, max_width):
+        """
+        Render region of a big image at an appropriate zoom level
+        so width < max_width
+        """
+
+        size_x = image.getSizeX()
+        size_y = image.getSizeY()
+        x = region['x']
+        y = region['y']
+        width = region['width']
+        height = region['height']
+
+        zm_levels = image.getZoomLevelScaling()
+        # e.g. {0: 1.0, 1: 0.25, 2: 0.0625, 3: 0.03123, 4: 0.01440}
+        # Pick zoom such that returned image is below MAX size
+        max_level = len(zm_levels.keys()) - 1
+
+        # Maximum size that the rendering engine will render without OOM
+        max_plane = self.conn.getDownloadAsMaxSizeSetting()
+
+        # start big, and go until we reach target size
+        zm = 0
+        while (zm <= max_level and
+               zm_levels[zm] * width > max_width or
+               zm_levels[zm] * width * zm_levels[zm] * height > max_plane):
+            zm = zm + 1
+
+        level = max_level - zm
+
+        # We need to use final rendered jpeg coordinates
+        # Convert from original image coordinates by scaling
+        scale = zm_levels[zm]
+        x = int(x * scale)
+        y = int(y * scale)
+        width = int(width * scale)
+        height = int(height * scale)
+        size_x = int(size_x * scale)
+        size_y = int(size_y * scale)
+
+        canvas = None
+        # Coordinates below are all final jpeg coordinates & sizes
+        if x < 0 or y < 0 or (x + width) > size_x or (y + height) > size_y:
+            # If we're outside the bounds of the image...
+            # Need to render reduced region and paste on to full size image
+            canvas = Image.new("RGBA", (width, height), (221, 221, 221))
+            paste_x = 0
+            paste_y = 0
+            if x < 0:
+                paste_x = -x
+                width = width + x
+                x = 0
+            if y < 0:
+                paste_y = -y
+                height = height + y
+                y = 0
+
+        # Render the region...
+        jpeg_data = image.renderJpegRegion(z, t, x, y, width, height,
```

Comment:
handle ``jpeg_data = None``

Reply:
Done in https://github.com/ome/omero-figure/pull/243/commits/d4325e19f6bacc26058c0d411a39fc960d1213a2